src/main/java/de/tum/in/www1/artemis/service/programming/ProgrammingExerciseGradingService.java

```diff
@@ -571,97 +599,189 @@ private boolean createFeedbackForDuplicateTests(Result result, ProgrammingExerci
+        successfulTestPoints -= calculateTotalPenalty(programmingExercise, result.getParticipation(), staticCodeAnalysisFeedback, applySubmissionPolicy);
+
+        if (successfulTestPoints < 0) {
+            successfulTestPoints = 0;
+        }
 
-            // The score is calculated as a percentage of the maximum points
-            double score = successfulTestPoints / programmingExercise.getMaxPoints() * 100.0;
+        // The score is calculated as a percentage of the maximum points
+        return successfulTestPoints / programmingExercise.getMaxPoints() * 100.0;
+    }
 
-            result.setScore(score);
+    /**
+     * Calculates the total points that should be given for the successful test cases.
+     *
+     * Additionally, updates the feedback in the result for each passed test case with the points
+     * received for that specific test case.
+     *
+     * Does not apply any penalties to the score yet.
+     *
+     * @param programmingExercise which the result belongs to.
+     * @param result for which the points should be calculated.
+     * @param successfulTestCases all test cases the submission passed.
+     * @param totalTestCaseCount the total number of relevant test cases. This might not be the total
+     *                           number of test cases in the exercise as some test cases are ignored
+     *                           for the calculation before the exercise due date.
+     * @param weightSum the sum of test case weights of all test cases that have to be considered.
+     * @return the total score for this result without penalty deductions.
+     */
+    private double calculateSuccessfulTestPoints(final ProgrammingExercise programmingExercise, final Result result, final Set<ProgrammingExerciseTestCase> successfulTestCases,
+            int totalTestCaseCount, double weightSum) {
+        double successfulTestPoints = successfulTestCases.stream().mapToDouble(test -> {
+            double credits = calculatePointsForSuccessfulTestCase(result, programmingExercise, test, totalTestCaseCount, weightSum);
+            setCreditsForTestCaseFeedback(result, test, credits);
+            return credits;
+        }).sum();
+
+        /*
+         * The points are capped by the maximum achievable points. The cap is applied before the static code analysis penalty is subtracted as otherwise the penalty won't have any
+         * effect in some cases. For example with maxPoints=20, successfulTestPoints=30 and penalty=10, a student would still receive the full 20 points, if the points are not
+         * capped before the penalty is subtracted. With the implemented order in place successfulTestPoints will be capped to 20 points first, then the penalty is subtracted
+         * resulting in 10 points.
+         */
+        double maxPoints = programmingExercise.getMaxPoints() + Optional.ofNullable(programmingExercise.getBonusPoints()).orElse(0.0);
```

Comment:
This could be another method as it introduces a new functionality (capping the points)

Reply:
I agree, it is now in its own method.